File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        
        def backtrack(start, curr_sum, path):
            if curr_sum == target:
                res.append(path)
                return
            if curr_sum > target:
                return

            for i in range(start, len(candidates)):
                if i > start and candidates[i] == candidates[i-1]:
                    continue
                backtrack(i+1, curr_sum+candidates[i], path+[candidates[i]])
        
        backtrack(0, 0, [])
        return res
    
            
            

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        helper(candidates, target, 0, ans, {});
        return ans;
    }

private :
    void helper(vector<int>& c, int t, int i, vector<vector<int>> &ans, vector<int> curr){
        if(t==0){
            ans.push_back(curr);
            return ;
        }
        for(int j = i; j<c.size(); j++){
            if(j>i && c[j]==c[j-1]) continue;
            if(c[j] > t) break;

            curr.push_back(c[j]);
            helper(c,t-c[j], j+1, ans, curr);
            curr.pop_back();
        }
    }
};